.github/workflows/ci.yml

```diff
@@ -0,0 +1,139 @@
+name: CI
+
+on:
+    push:
+        branches:
+            - "main"
+            - "release/*"
+    pull_request:
+        branches:
+            - "main"
+            - "release/*"
+    release:
+        types: [created]
+
+jobs:
+    lint:
+        name: Lint Python
+        runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                python-version: ["3.10", "3.11", "3.12"]
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4
+
+            - name: Cache dependencies
+              uses: actions/cache@v4
+              with:
+                path: ~/.cache/pip
+                key: ${{ runner.os }}-pip-${{ hashFiles('**/pyproject.toml') }}
+                restore-keys: |
+                  ${{ runner.os }}-pip-
+
+            - name: Set up Python
+              uses: actions/setup-python@v5
+              with:
+                python-version: ${{ matrix.python-version }}
+
+            - name: Install dependencies
+              run: pip install .[dev]
+
+            - name: Run Linting
+              run: ruff check src/ tests/
+
+    mypy:
+        name: Mypy
+        runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                python-version: ["3.10", "3.11", "3.12"]
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4
+
+            - name: Cache dependencies
+              uses: actions/cache@v4
+              with:
+                path: ~/.cache/pip
+                key: ${{ runner.os }}-pip-${{ hashFiles('**/pyproject.toml') }}
+                restore-keys: |
+                  ${{ runner.os }}-pip-
+
+            - name: Set up Python
+              uses: actions/setup-python@v5
+              with:
+                python-version: ${{ matrix.python-version }}
+
+            - name: Install dependencies
+              run: pip install .[dev]
+
+            - name: Run Mypy
+              run: mypy src/ tests/
```

Comment:
We also have `ruff` and `mypy` in the pre-commit, it is better to catch these errors before we push the repo I think.

Reply:
Yes I have set up both intentionally, because there is technically nothing stopping someone from explicitly ignoring pre-commit when making their commits. But agree we should not be doing that as a rule of thumb 😄